### PR TITLE
Add fillcombinations function

### PR DESCRIPTION
**Pull request by bkamins**
Update of https://github.com/JuliaData/DataFrames.jl/pull/1864.

### CODE REVIEW

**Comment by nalimilan** [comment]
I think the logic is fine, though it's not easy to check without a docstring or tests. :-)

I'm not sure about the name. It's true that `expand` might be too general. OTOH that term doesn't seem to be used by any packages currently, and I'm not sure what else it could mean. `expandlevels` isn't totally accurate as the function doesn't just expand levels, it adds rows for missing combinations.

**Comment by bkamins** [comment]
@nalimilan I have added tests. The PR is good for review now.
The bad thing is that it is quite complex in corner cases and essentially every test I add is relevant and checks something important.

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
...
+                      source=string("source", idx_ind),
+                      matchmissing=:equal)
+        end
+        # Replace missing values with the fill
+        if !ismissing(fill)
+            mask = out_df[!, end] .== "left_only"
+            if count(mask) > 0
+                for n in length(colind)+1:ncol(out_df)-1
+                    tmp_col = out_df[!, n]
+                    out_col = similar(tmp_col,
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:1584]: @nalimilan - this is needed to have consistent behavior with `unstack`, see also https://github.com/JuliaData/CategoricalArrays.jl/issues/380.

File: src/abstractdataframe/abstractdataframe.jl [hunk 2]
```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
...
+    If `allcols` is `true` all columns from `df` are added to the resulting data
+    frame (the `indexcols` are the first columns stored in the result). For the
+    combinations of `indexcols` these columns take `fill` value (`missing`) by
+    default.
+
+    If `allowduplicates` is `false` (the default) `indexcols` may only contain
+    unique combinations of `indexcols` values. If `allowduplicates` is `true`
+    duplicates are allowed. They are not repeated if `allcols` is `false`
+    only unique combinations are produced then, but if `allcols` is `true`
+    the duplicates are included.
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1472]: This behavior differs from dplyr's `complete`, which retains duplicates. Any particular reason to drop them? Or should we allow three options about how duplicates should be handled?

File: src/abstractdataframe/abstractdataframe.jl [hunk 3]
```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
...
+        out_df[!, _names(df)[cind]] = repeat(target_col, inner=last_inner, outer=outer)
+    end
+    @assert inner == target_rows
+
+    # optionally join the remaining columns
+    if allcols
+        idx_ind = 0
+        while columnindex(df, string("source", idx_ind)) != 0
+            idx_ind += 1
+        end
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1562]: These uncovered lines appear to be real?

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
...
+    end
+
+    # construct expanded columns
+    out_df = DataFrame()
+    inner = 1
+    @assert length(uniqueVals) == length(colind)
+    for (val, cind) in zip(uniqueVals, colind)
+        len = length(val)
+        target_col = Tables.allocatecolumn(eltype(df[!, cind]), len)
+        target_col .= val
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:1548]: It is the opposite `val` is a vector of the same length as `target_col`, so `fill!` is incorrect. I will switch to `copyto!` though.

File: test/data.jl [hunk 5]
```diff
@@ -473,4 +477,106 @@ end
...
+        @test isequal_coltyped(completecombinations(df1, [:a, :b], allcols=true, allowduplicates=ad),
+                               DataFrame(a=[1, 2, missing, 1, 2, missing],
+                                         b=[1, 1, 1, 2, 2, 2],
+                                         c=categorical([11, 12, missing, missing, missing, missing]),
+                                         d=[111, 112, missing, missing, missing, 113]))
+        @test isequal_coltyped(completecombinations(df1, [:a, :b], allcols=true, fill="X", allowduplicates=ad),
+                               DataFrame(a=[1, 2, missing, 1, 2, missing],
+                                         b=[1, 1, 1, 2, 2, 2],
+                                         c=[11, 12, "X", "X", "X", missing],
+                                         d=[111, 112, "X", "X", "X", 113]))
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/data.jl:505]: Yes - I made this choice intentionally. The same behavior of `fill` is in `unstack`.

File: src/abstractdataframe/abstractdataframe.jl [hunk 6]
```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
...
+        throw(ArgumentError("duplicate rows in input"))
+    end
+
+    # Create a vector of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        # All levels are retained, missing is added only if present
+        if any(ismissing, df[!, col])
+            tempcol = vcat(levels(df[!, col]), missing)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:1531]: I have opened https://github.com/JuliaData/DataAPI.jl/issues/44

File: src/abstractdataframe/abstractdataframe.jl [hunk 7]
```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
...
+        else
+            tempcol = levels(df[!, col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # make sure we do not overflow in the target data frame size
+    target_rows = Int(foldl((x, y) -> x * length(y), uniqueVals, init=big(1)))
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:1539]: this is OK

File: src/abstractdataframe/abstractdataframe.jl (hunk 2 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: It is a consequence of the current design. I.e. it is natural to drop them in the way it is implemented.
But we can discuss what to do.

Now I am thinking that maybe we do not need the `allcols` kwarg and can simplify the API. Maybe we should have `allcols=true` always and expect from the user to pass a data frame only with columns that are wanted in the output?

File: test/data.jl [hunk 8]
```diff
@@ -473,4 +477,106 @@ end
...
+
+    for ad in (true, false)
+        # allcols=false
+        @test isequal_coltyped(completecombinations(df1, [:a, :b], allowduplicates=ad),
+                               DataFrame(a=[1, 2, missing, 1, 2, missing],
+                                         b=[1, 1, 1, 2, 2, 2]))
+        @test isequal_coltyped(completecombinations(df1, :c, allowduplicates=ad),
+                               DataFrame(c=categorical([12, 11, 10, missing])))
+        @test isequal_coltyped(completecombinations(df1, [:c, :b], allowduplicates=ad),
+                               DataFrame(c=categorical([12, 11, 10, missing, 12, 11, 10, missing]),
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/data.jl:493]: done

File: src/abstractdataframe/abstractdataframe.jl (hunk 3 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:1566]: this is covered.

**Comment by bkamins** [comment]
`allcols` kwarg is now removed.

If duplicates are allowed they are always kept.

also since we always keep all cols the column order of the source data frame is retained (earlier I put the expanded columns first, but since now we always keep all columns I think it makes more sense to keep their original order).

This should be good for a review.

**Comment by bkamins** [comment]
@nalimilan - no rush, but it would be good to finalize this PR before we forget what we wanted. Thank you!

File: src/abstractdataframe/abstractdataframe.jl [hunk 9]
```diff
@@ -1450,6 +1450,146 @@ julia> unique!(df)  # modifies df
...
+"""
+    completecombinations(df::AbstractDataFrame, indexcols;
+                         allowduplicates::Bool=false,
+                         fill=missing)
+
+Generate all combinations of levels of column(s) `indexcols` in data frame `df`.
+Levels and their order are determined by the `levels` function
+(i.e. unique values sorted lexicographically by default, or a custom set
+of levels for e.g. `CategoricalArray` columns), in addition to `missing` if present.
+
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1462]: This still applies, right?
```suggestion

All columns from `df` are added at the end of the returned data frame.
For combinations of `indexcols` not present in `df` these columns are
filled with the `fill` value (`missing` by default).

```

File: src/abstractdataframe/abstractdataframe.jl [hunk 10]
```diff
@@ -1450,6 +1450,146 @@ julia> unique!(df)  # modifies df
...
+If `allowduplicates=false` (the default) `indexcols` may only contain
+unique combinations of `indexcols` values. If `allowduplicates=true`
+duplicates are allowed
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1465]:
```suggestion
duplicates are allowed.
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 11]
```diff
@@ -1450,6 +1450,146 @@ julia> unique!(df)  # modifies df
...
+
+    if has_duplicates
+        order_ind = 0
+        while columnindex(df, string("order9427", order_ind)) != 0
+            order_ind += 1
+        end
+        insertcols!(out_df, 1, string("order9427", order_ind) => 1:nrow(out_df))
+        out_df = leftjoin(out_df, df; on=_names(df)[colind],
+                            source=string("source7249", idx_ind),
+                            matchmissing=:equal)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1565]:
```suggestion
        out_df = leftjoin(out_df, df; on=_names(df)[colind],
                          source=string("source7249", idx_ind),
                          matchmissing=:equal)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 12]
```diff
@@ -1450,6 +1450,146 @@ julia> unique!(df)  # modifies df
...
+
+    # Replace missing values with the fill
+    if !ismissing(fill)
+        mask = out_df[!, end] .== "left_only"
+        if count(mask) > 0
+            for n in length(colind)+1:ncol(out_df)-1
+                tmp_col = out_df[!, n]
+                out_col = similar(tmp_col,
+                                    promote_type(eltype(tmp_col), typeof(fill)),
+                                    length(tmp_col))
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1582]:
```suggestion
                out_col = similar(tmp_col,
                                  promote_type(eltype(tmp_col), typeof(fill)),
                                  length(tmp_col))
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 13]
```diff
@@ -1450,6 +1450,146 @@ julia> unique!(df)  # modifies df
...
+        insertcols!(out_df, 1, string("order9427", order_ind) => 1:nrow(out_df))
+        out_df = leftjoin(out_df, df; on=_names(df)[colind],
+                            source=string("source7249", idx_ind),
+                            matchmissing=:equal)
+        sort!(out_df, 1)
+        select!(out_df, 2:ncol(out_df))
+    else
+        leftjoin!(out_df, df; on=_names(df)[colind],
+                    source=string("source7249", idx_ind),
+                    matchmissing=:equal)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1571]:
```suggestion
        leftjoin!(out_df, df; on=_names(df)[colind],
                  source=string("source7249", idx_ind),
                  matchmissing=:equal)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 14]
```diff
@@ -1450,6 +1450,146 @@ julia> unique!(df)  # modifies df
...
+        outer, remv = divrem(target_rows, inner)
+        @assert iszero(remv)
+        out_df[!, _names(df)[cind]] = repeat(target_col, inner=last_inner, outer=outer)
+    end
+    @assert inner == target_rows
+
+    idx_ind = 0
+    while columnindex(df, string("source7249", idx_ind)) != 0
+        idx_ind += 1
+    end
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:1555]: Maybe storing the column name in a variable and using it later instead of calling `string("source7249", idx_ind)` would be slightly cleaner? Something like
```suggestion
    idx_ind = 0
    while true
        idx_col = string("source7249", idx_ind)
        columnindex(df, idx_col) == 0 && break
        idx_ind += 1
    end
```

Same for `order_ind`.

File: test/data.jl [hunk 15]
```diff
@@ -473,4 +477,109 @@ end
...
+        @test isequal_coltyped(completecombinations(df1, :c, fill="X", allowduplicates=ad),
+                               DataFrame(a=[2, 1, "X", missing],
+                                         b=[1, 1, "X", 2],
+                                         c=categorical([12, 11, 10, missing]),
+                                         d=[112, 111, "X", 113]))
+        @test isequal_coltyped(completecombinations(df1, [:c, :b], allowduplicates=ad),
+                               DataFrame(a=[2; 1; fill(missing, 6)],
+                                         b=[1, 1, 1, 1, 2, 2, 2, 2],
+                                         c=categorical([12, 11, 10, missing, 12, 11, 10, missing]),
+                                         d=[112; 111; fill(missing, 5); 113]))
```
Comment (nalimilan) [inline test/data.jl:510]: I don't see the test for `levels`. Am I missing it?

**Review by nalimilan — COMMENTED** [review]
Looks good! Just minor comments.

Maybe we should check that the name sounds OK by asking people on Slack?

File: src/abstractdataframe/abstractdataframe.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply src/abstractdataframe/abstractdataframe.jl:1522]: https://github.com/JuliaData/DataAPI.jl/pull/46

File: src/abstractdataframe/abstractdataframe.jl (hunk 9 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:1462]: I have corrected your suggestion this. Now we keep the columns in the order of source data frame (but indeed I have accidentally deleted the `fill` part).

File: test/data.jl (hunk 15 above)
Comment (bkamins) [reply test/data.jl]: I have added the tests - as usual tricky cases are present.

**Comment by bkamins** [comment]
Proposal for the name from Slack I like best is `fillcombinations`.

Other: `add_combinations`, `fill_combinations`, `complete_combinations`

**Comment by bkamins** [comment]
@nalimilan - so what do you think about the best name for the function given the discussion?

**Comment by nalimilan** [comment]
I also prefer `fillcombinations`, though `completecombinations` is OK too.

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!
